packages/funding-rate-proposer/src/proposer.js

```diff
@@ -0,0 +1,360 @@
+    };
+    this.createConfigStoreContract = storeAddress => {
+      return new this.web3.eth.Contract(getAbi("ConfigStore"), storeAddress);
+    };
+
+    // Gas Estimator to calculate the current Fast gas rate.
+    this.gasEstimator = gasEstimator;
+
+    // Cached mapping of identifiers to pricefeed classes.
+    this.priceFeedCache = {};
+
+    // Cached perpetual contracts and precomputed state.
+    this.contractCache = {};
+
+    // Helper functions from web3.
+    this.BN = this.web3.utils.BN;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+
+    // Default config settings. Bot deployer can override these settings by passing in new
+    // values via the `perpetualProposerConfig` input object. The `isValid` property is a function that should be called
+    // before resetting any config settings. `isValid` must return a Boolean.
+    const defaultConfig = {
+      fundingRateErrorPercent: {
+        //   "fundingRateErrorPercent":0.05 ->  Current funding rates (as stored in the OptimisticOracle)
+        //                                      that do not equal the bot's queried funding rate
+        //                                      within this error % will be requested to be updated,
+        //                                      and proposed to.
+        //                                      e.g. 0.05 implies 5% margin of error.
+        value: 0.05,
+        isValid: x => {
+          return !isNaN(x);
+          // Negative allowed-margins might be useful based on the implementation
+          // of `isDeviationOutsideErrorMargin()`
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(perpetualProposerConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
+    await Promise.all([this.perpetualFactoryClient.update(), this.gasEstimator.update()]);
+
+    // Once PerpFactory client is updated, cache contract instances for each address deployed.
+    // These will be saved to this.contractCache. Additionally, precompute state that we'll
+    // use in `updateFundingRates()` such as the latest funding rates and config store settings.
+    await this._cachePerpetualContracts();
+
+    // The following updates can be done independently after contract state is fetched:
+    await Promise.all([
+      // Increase allowances for all contracts to spend the bot owner's respective collateral currency.
+      this._setAllowances(),
+      // For each contract, create, save, and update a pricefeed instance for its identifier,
+      // or just update the pricefeed if the identifier already has cached an instance.
+      await this._cacheAndUpdatePriceFeeds()
```

Comment:
```suggestion
      this._cacheAndUpdatePriceFeeds()
```

Reply:
This is an async method because `createRefernecePriceFeed...` is async

Reply:
But it's being called inside a `Promise.all`, no? _setAllowances above is also async, but you don't put an await in front of it because the promise.all handles that if I'm reading this correctly.

Reply:
oh true, done

Reply:
Bump.

Reply:
This is done.